### ISSUE TITLE
Let users specify the Mesos role for each Marathon app

### ISSUE THREAD

**Issue by ConnorDoyle** [mesosphere/marathon#1460]
In the [future](https://issues.apache.org/jira/browse/MESOS-1763), frameworks will be able to register with multiple roles.  When this happens, we'll want the ability to tell Marathon which role's resources to use to launch tasks for each app.

Frameworks are currently offered resources associated both with the registration role and the default (`*`) role.  We already allow users to choose what role to register as (via the `--mesos_role` flag).  The current behavior is to ignore the role associated with each resource offer when launching tasks.  A per-app role would already be useful to distinguish between these sets of resources.

Currently, and even more so in the "future" scenario, a `--default_roles` flag could tune the scheduler behavior when inspecting offer roles.  For example given that a Marathon scheduler is registered with role `A`, receiving offers assiciated with roles `A` and `*`, the administrator may want to control the default roles for apps that do not specify per-app roles.

cc @cmaloney @pyronicide

**To summarize the proposal:**

- [ ] Introduce a `--default_app_roles` command line flag to Marathon that defaults to `*` if no value for `--mesos_role` is supplied and `*,A` if `--mesos_role A` is supplied.
- [ ] Introduce a `mesosRoles` (or similarly named) attribute to the App definition to control the role-associated resources to use for launching tasks for that app.  Validation should ensure that the value is either `*` or the role that Marathon used to register with Mesos.

**Comment by gomes** [comment]
@ConnorDoyle MESOS-1763 seems not to be a priority, as @tnachen said, do you plan to focus on it?

**Comment by kolloch** [comment]
Can you elaborate on the use case for that, please?

Thats what I understood from what @adam-mesos was writing to us:

Context:
* you want to mark certain slaves hosts as public
* you create a special "public" marathon instance of these hosts
* you might have any number of other marathon instances

You need the following properties:
A) tasks created by other marathons (non "public" ones) will never be launched on the public hosts, either by accident or on purpose
B) you can launch tasks on the public hosts via the special marathon instance. That might be the default for that marathon or you will have to specify that role explicitly in the app definition.
C) you do not want to have a separate mesos cluster for the public nodes but reuse masters/zookeepers

B) and C) can be achieved by using attributes and the associated constraints. A) is the difficult and new part. Unfortunately, I have not a good understanding of the mesos roles idea yet, so I am not able to judge if that's the concept you would want to use for this.

**Comment by pyronicide** [comment]
Here's a summary of a chat we had:

The use case is that we need a way to do external -> internal routing to marathon apps. To manage this with mesos, we're introducing a new class of slaves that have a "public" role. Those slaves will have external ip addresses and some ports open. The public role will ensure that anyone who hasn't registered under that role cannot schedule on those slaves.

We need marathon to be aware of roles, so that when it registers as "public", we can create an app definition that will only ever be scheduled on a slave with a "public" role. All existing app definitions will be guaranteed to *never* be scheduled on a "public" slave.

**Comment by kolloch** [comment]
Ok, I guess that we can solve this by adding a "resourceRole" attribute to app definitions. If unset, tasks will be launched on resource offers from mesos of any role. If set, only resources with that role will be considered for launching the tasks.

**Comment by kolloch** [comment]
@pyronicide's comment that I deleted by accident:

Here's a summary of a chat we had:

The use case is that we need a way to do external -> internal routing to marathon apps. To manage this with mesos, we're introducing a new class of slaves that have a "public" role. Those slaves will have external ip addresses and some ports open. The public role will ensure that anyone who hasn't registered under that role cannot schedule on those slaves.

We need marathon to be aware of roles, so that when it registers as "public", we can create an app definition that will only ever be scheduled on a slave with a "public" role. All existing app definitions will be guaranteed to never be scheduled on a "public" slave.

**Comment by kolloch** [comment]
If I understand it correctly, mesos will only offer "*" resources to a Marathon instance by default. You can choose the role of a Marathon instance with the "--mesos_role" CLI option, in that case you see "*" resources and resources of the specified role.

If you configured your slave to offer all its resources with role “public”, only frameworks that have the role “public” will see these resources. So, you do not have to worry about any unspecialized marathons launching tasks on “public”.

To launch tasks on “public”, you could launch a special marathon instance registered with the role “public”. To prevent it from starting your apps on "*" resources you could use additional [attribute constraints](https://github.com/mesosphere/marathon/blob/master/docs/docs/constraints.md#attribute-field). You just have to make sure that you can identify your "public" slaves by attributes as well, for example by specifying `--attributes='network:public'` for that slave and using the corresponding constraint, e.g. "network:LIKE:public".

**Comment by ConnorDoyle** [comment]
@kolloch one drawback to an attributes-based solution is that it isn't enforceable via the built-in framework authentication mechanism as is possible with roles.

**Comment by gomes** [comment]
Another user case (that is my case right now):
I am using Qubit Bamboo as my service discovery, I run one instance in every slave I have, so my apps access only localhost. The communications between apps are always using the service discovery. I have to assure that port 8080 is free to launch the service discovery, but Mesos usually offer 8080 to another apps. I'm changing my actual config to two marathons, one will use resources(*) with the usual 31000-32000 ports and the another one setting the role with the 8080 port available

**Comment by gomes** [comment]
I think https://github.com/mesosphere/marathon/issues/1015 can be solved using roles too.

**Comment by adam-mesos** [comment]
> an attributes-based solution ... isn't enforceable via the built-in framework authentication mechanism as is possible with roles.
It's true that Mesos already provides an ACL for what roles a particular framework principal registers as, but we can add more ACLs as necessary. There is currently no ACL stating which roles a particular slave principal is allowed to reserve resources for, not which attribute keys a slave can use. But we could add either of those. Since attributes are not used by resource accounting, and are a part of every resource offer, I see no reason (yet?) why we'd want to prevent certain framework principals from seeing a particular attribute.

I think that the public/private node use case can already be solved with a combination of reserved resources for the public role to prevent (enforceable via ACLs) non-public frameworks from getting public resources, and attributes-based constraints (hand-set by app admin) to prevent the public Marathon from launching these special apps on nodes/resources not labelled public. I don't see an urgent need for this feature based on that use case.

gomes' use case seems like it can be solved with a framework-level role reservation too, so I don't think it calls for app-level roles either.

**Comment by kolloch** [comment]
I am not completely sold on the vision of using a set of DMZ nodes (edge routers) with Mesos.

Yet, I think that this feature could be useful in other use cases as well. A user might want to designate a share of resources for the role "production" to ensure that other tasks (e.g. test tasks) do not consume all resources. Untouched frameworks should never launch tasks on resources of role "production" (that is how it works today). Additionally, if we use one Marathon both for production and test tasks, the default should be to launch tasks on resources other than of role "production" (not possible today). Only the tasks of an app that is specifically designated as production should run on role "production" (not possible today).

## Connor's approach (slightly adapted)

@ConnorDoyle's proposal would also support the use cases above. I reiterate that approach with examples and slightly different field/CLI option names below:

* Introduce an optional attribute `acceptedResourceRoles` in the app definition which allows an array of role names as strings. All resources which have any of the specified roles are considered for launching tasks of that app. Resources with other roles are ignored.
* We also would allow a global Marathon-wide default for the `acceptedResourceRoles` attribute with `--default_accepted_resource_roles` as a comma-separated role list. If `--mesos_role A` is used, we default to `--default_accepted_resource_roles=*,A` to preserve the current behavior. If `--mesos_role` is not used, we default to `--default_accepted_resource_roles=*`. 

### Application to use cases

In the edge routing use case with public nodes we can use `"acceptedResourceRoles": ["public"]` for publicly visible apps and `"acceptedResourceRoles": ["*"]` for private apps. We probably wouldn't specify the latter explicitly but use a Marathon-wide default of `--default_accepted_resource_roles=*`.

In the production resource use case we can use `"acceptedResourceRoles": ["production", "*"]` for production apps and use the default of `--default_accepted_resource_roles=\*` for everything else.

### Validation

Both the global default and the app definition attribute should be validated. The `--default_accepted_resource_roles` setting is validated at startup. Only "\*" and a potentially specified `--mesos_role` are allowed. The `"acceptedResourceRoles"` attribute is validated on every app definition change. An empty array is rejected because that would mean that no offer would ever match. Otherwise we allow only roles for which we potentially see resource offers. Currently that's "\*" and the role that the user specified by `--mesos_role` (if any).

## Alternative constraint-based approach

A related concept that we have in Marathon are constraints. These work on attributes and the hostname. We could extend that to roles. I discussed that today with colleagues intensely, @lloesche & @aquamatthias, yesterday with @drexin. If a user does not supply constraints or doesn't explicitly provide a role constraint, we would to need to add this default constraint. Unfortunately, this approach feels awkward because you do not only narrow down the choices but you actually want to use this to segment the cluster. Therefore, without further ideas, we rejected that approach.

## Label-based approach

@lloesche quite likes the usage of labels in the google cluster definitions. So instead of directly coupling app definitions to roles, you could use rules to do that. These rules would assign accepted resource roles based on the labels. This approach hinges on whether we could come up with a good rule language for that.

**Comment by solarkennedy** [comment]
It sounds like the design of this only allows *1* `--mesos_role`, so you can kinda get your `*` or `public`, but you only get those two?

Mesos allows a csv of roles to have registered with the master. It sounds like individual frameworks can only register *1* other role? :(

**Comment by cmaloney** [comment]
Mesos only allows frameworks to have one role currently (See: https://issues.apache.org/jira/browse/MESOS-1763). So you can give just the one named role. The "*" resource in Mesos is the general / common resource pool which every resource has access to.

**Comment by JamesWiley** [comment]
This doesn't work.  Tried specifying a mesos-role for Marathon that was responsible for launching docker images associated with that one role and the images would never come up.  Once the role specifier was removed from Marathon and the mesos-slaves, everything launched fine.

**Comment by neerajgoel82** [comment]
As per my understanding, we can start marathon with single role only (specified as mesos_role flag). Can we supply multiple mesos_roles at the beginning like role1, role2 ... roleN and specify each job to be run as one of these roles?

**Comment by adam-mesos** [comment]
The first words of this ticket are "In the future," which links to https://issues.apache.org/jira/browse/MESOS-1763 which is still unresolved.
Latest news is that we're investigating hierarchical roles in addition to just a flat list per framework.

**Comment by neerajgoel82** [comment]
Sorry I missed the link. 

Thanks a lot for sharing the update. Looking forward to this feature.
